api/api.go

```diff
@@ -0,0 +1,443 @@
+	req.Header.Set("Accept", "application/json")
+
+	client := &http.Client{Timeout: 20 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != expectStatus {
+		return fmt.Errorf("request failed: %d - %s", resp.StatusCode, resp.Status)
+	}
+
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(b, &v)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Version get version of dockerd server
+func (api *API) version() (string, error) {
+	var res dockerTypes.Version
+	if err := api.get("/version", "", &res); err != nil {
+		return "", err
+	}
+	return res.APIVersion, nil
+}
+
+// Build build a project
+func (api *API) Build(project types.Project) (types.Service, error) {
+	dir, err := ioutil.TempDir("/tmp", "fx-build-dir")
+	if err != nil {
+		return types.Service{}, err
+	}
+
+	defer os.RemoveAll(dir)
+
+	for _, file := range project.Files {
+		tmpfn := filepath.Join(dir, file.Path)
+		if err := ioutil.WriteFile(tmpfn, []byte(file.Body), 0666); err != nil {
+			return types.Service{}, err
+		}
+	}
+
+	tarDir, err := ioutil.TempDir("/tmp", "fx-tar")
+	if err != nil {
+		return types.Service{}, err
+	}
+
+	imageID := uuid.New().String()
+	tarFilePath := filepath.Join(tarDir, fmt.Sprintf("%s.tar", imageID))
+	if err := utils.TarDir(dir, tarFilePath); err != nil {
+		return types.Service{}, err
+	}
+
+	dockerBuildContext, err := os.Open(tarFilePath)
+	if err != nil {
+		return types.Service{}, err
+	}
+	defer dockerBuildContext.Close()
+
+	type buildQuery struct {
+		Labels     string `url:"labels"`
+		Tags       string `url:"t"`
+		Dockerfile string `url:"dockerfile"`
+	}
+
+	// Apply default labels
+	labelsJSON, _ := json.Marshal(
+		map[string]string{
+			"belong-to": "fx",
+		},
+	)
+
+	q := buildQuery{
+		Tags:       imageID,
+		Labels:     string(labelsJSON),
+		Dockerfile: "Dockerfile",
+	}
+	qs, err := query.Values(q)
+	if err != nil {
+		return types.Service{}, err
+	}
+
+	version, err := api.version()
+	if err != nil {
+		return types.Service{}, err
+	}
+	path := "/build"
+	url := fmt.Sprintf("http://%s/v%s%s?%s", api.endpoint, version, path, qs.Encode())
+	req, err := http.NewRequest("POST", url, dockerBuildContext)
+	if err != nil {
+		return types.Service{}, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-tar")
+	client := &http.Client{Timeout: 120 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		return types.Service{}, err
+	}
+
+	defer resp.Body.Close()
+	scanner := bufio.NewScanner(resp.Body)
+
+	for scanner.Scan() {
+		// TODO Maybe need log something out
+		// fmt.Printf("%s\n", scanner.Text())
+	}
+	if err := scanner.Err(); err != nil {
+		return types.Service{}, err
+	}
+
+	return types.Service{
+		Name:   project.Name,
+		Status: types.ServiceStatusInit,
+		Image:  imageID,
+	}, nil
+}
+
+// List list service
+func (api *API) list(name string) ([]types.Service, error) {
```

Comment:
We do have a command `fx list`, which is handled by `api.List`,  so maybe this `list` should be redesign.